### PR TITLE
Feat/add steep transport

### PR DESCRIPTION
**Pull request by chloelbn**
Adds type checking of Transport layer
Depends on https://github.com/algolia/algoliasearch-client-ruby/pull/401

### CODE REVIEW

File: lib/algolia/transport/request_options.rb [hunk 1]
```diff
@@ -34,7 +34,7 @@ def create(opts = {})
       def add_headers(opts = {})
         unless opts[:headers].nil?
           opts[:headers].each do |opt, value|
-            @headers[opt.to_sym] = value
+            @headers[opt.to_s] = value
```
Comment (DevinCodes) [inline lib/algolia/transport/request_options.rb:37]: Why do we cast to a string instead of a symbol?

File: lib/algolia/transport/transport.rb [hunk 2]
```diff
@@ -127,15 +127,15 @@ def build_body(body, request_options, method)
...
-      def generate_headers(request_options = {})
-        headers                                                     = {}
-        extra_headers                                               = request_options.headers || {}
-        @config.default_headers.each { |key, val| headers[key.to_s] = val }
-        extra_headers.each { |key, val| headers[key.to_s]           = val }
+      def generate_headers(request_options)
+        headers                                                = {}
+        extra_headers                                          = request_options.headers
+        @config.default_headers.each { |key, val| headers[key] = val }
+        extra_headers.each { |key, val| headers[key]           = val }
```
Comment (DevinCodes) [inline lib/algolia/transport/transport.rb:138]:
```suggestion
        headers = @config.default_headers.merge(request_options.headers)
```

File: sig/transport/request_options.rbs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  module Transport
+    class RequestOptions
+      attr_accessor headers: Hash[String, String]
+      attr_accessor params: Hash[Symbol, String]
+      attr_accessor data: Hash[Symbol, String]
+      attr_accessor timeout: nil|Integer
+      attr_accessor connect_timeout: nil|Integer
+      attr_accessor compression_type: String
+
+      def initialize: (Search::Config config) -> bool
```
Comment (DevinCodes) [inline sig/transport/request_options.rbs:11]:
```suggestion
      def initialize: (Search::Config config) -> void
```

File: sig/transport/request_options.rbs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+      # Create and format headers and params from request options
```
Comment (DevinCodes) [inline sig/transport/request_options.rbs:13]: As with your other PR, I think we can remove the comments in `.rbs` files 🙂

File: sig/transport/stateful_host.rbs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    class StatefulHost
+      include CallType
+      attr_reader url: String
+      attr_reader protocol: String
+      attr_reader accept: String
+      attr_accessor last_use: Time
+      attr_accessor retry_count: Integer
+      attr_accessor up: bool
+
+      def initialize: (String url, ?::Hash[Symbol, String|[String]] opts) -> bool
```
Comment (DevinCodes) [inline sig/transport/stateful_host.rbs:13]:
```suggestion
      def initialize: (String url, ?::Hash[Symbol, String|[String]] opts) -> void
```

File: sig/transport/transport.rbs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+module Algolia
+  module Transport
+    class Transport
+      include RetryOutcomeType
+      include CallType
+      #TODO include Helpers
+      include Helpers
+
+      def initialize: (Search::Config config, Object requester) -> bool
```
Comment (DevinCodes) [inline sig/transport/transport.rbs:9]:
```suggestion
      def initialize: (Search::Config config, Object requester) -> void
```

**Review by DevinCodes — CHANGES_REQUESTED** [review]

File: lib/algolia/transport/request_options.rb (hunk 1 above)
Comment (chloelbn) [reply lib/algolia/transport/request_options.rb:37]: The format of the headers will always be string anyway, because of headers like 'X-Algolia-XXX'

Comment (DevinCodes) [reply lib/algolia/transport/request_options.rb:37]: If it's always a string, we don't need to add the `.to_s` right? 🙂

Comment (chloelbn) [reply lib/algolia/transport/request_options.rb:37]: I meant they're supposed to be always string, so it's a way to ensure it's the case

File: sig/transport/request_options.rbs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+module Algolia
+  module Transport
+    class RequestOptions
+      attr_accessor headers: Hash[String, String]
+      attr_accessor params: Hash[Symbol, String]
```
Comment (DevinCodes) [inline sig/transport/request_options.rbs:5]: Can't the param values be a `bool` or `Integer` too?

File: sig/transport/transport.rbs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+module Algolia
+  module Transport
+    class Transport
+      include RetryOutcomeType
+      include CallType
+      #TODO include Helpers
```
Comment (DevinCodes) [inline sig/transport/transport.rbs:6]:
```suggestion
```

**Review by DevinCodes — CHANGES_REQUESTED** [review]

**Review by DevinCodes — APPROVED** [review]

**Pull request merged**